src/net/sourceforge/kolmafia/textui/langserver/FilesMonitor.java

```diff
@@ -0,0 +1,103 @@
+package net.sourceforge.kolmafia.textui.langserver;
+
+import java.io.File;
+import java.util.Arrays;
+import java.util.LinkedList;
+import java.util.List;
+import net.java.dev.spellcast.utilities.DataUtilities;
+import net.sourceforge.kolmafia.KoLConstants;
+
+/**
+ * Module in charge of the operations regarding our files, {@link Script scripts} and {@link
+ * Script.Handler handlers}, such as finding them, making them or traversing them.
+ */
+public class FilesMonitor {
+  final AshLanguageServer parent;
+
+  FilesMonitor(final AshLanguageServer parent) {
+    this.parent = parent;
+  }
+
+  /** Fetches or makes a Script for the given file. */
+  private Script getScript(final File file) {
+    synchronized (this.parent.scripts) {
+      Script script = this.parent.scripts.get(file);
+      if (script == null) {
+        script = new Script(this.parent, file);
+        this.parent.scripts.put(file, script);
+      }
+
+      return script;
+    }
+  }
+
+  /** Returns the currently existing {@link Script.Handler handler(s)} for the given file. */
+  List<Script.Handler> findHandlers(final File file) {
+    final List<Script.Handler> handlers = new LinkedList<>();
+
+    synchronized (this.parent.scripts) {
+      for (final Script script : this.parent.scripts.values()) {
+        if (script.handler != null
+            && script.handler.parser != null
+            && script.handler.parser.getImports().containsKey(file)) {
+          handlers.add(script.handler);
```

Comment:
IntelliJ tells me this line is uncovered. If true, could you add a test for successfully finding a handler?

Reply:
Mhh... this was mostly included "just in case". It will be formally used in the next step, where the server will start being able to update scripts' content.

Reply:
I am happy for this method to be a stub for now with something like `return new LinkedList<>()`, and a comment saying the real implementation is coming in the next step, if it will be easier to write a test in the next step.

But if it is coming in this step, I would rather a test be written here, so we don't forget about it.

Reply:
The thing is: it *is* currently possible to reach this line, but it is only possible as some sort of race condition.

Given a very big folder containing a *lot* of scripts.
`FilesMonitor.scan()` will iterate over its files and make sure there's a Handler for each of them.

If one of the very *first* script was to import one of the very *last* script, it would be possible for a Handler to exist for that script before `FilesMonitor.scan()` gets to it.

The thing is, due to the speed at which `FilesMonitor.scan()` operates, there would need to be a ***LOT*** of scripts for this to happen, so testing it would require cluttering `test/root`, a lot. Probably hundreds of files.

Reply:
Will you be able to add a test for this in one of the upcoming steps, or is it untestable in totality? If it will be testable later, add a `// TODO: test (step X)` and that'll be fine by me.

Is it testable as a non-integration test, if you mock out enough things?

Reply:
This line will end up getting tested in *the next* step. Right after this PR is accepted.

Reply:
Okay, great. Please add a comment along the lines of:
```
// TODO: currently untested, will be tested after server can update scripts' content
```
and then remove the comment in the upcoming PR, after the line is tested.

Reply:
...could I please not?
As you can see below, there's currently some jankiness going on with tests (unrelated to this PR) making them fail inconsistently, and now that I finally got them to work, I'd prefer to not jeopardize them again just for a comment :-(

Reply:
If a comment breaks the tests, they are broken, and should be fixed.

If the tests don't consistently pass, they are broken, and should be fixed.

If this PR gets merged with tests that fail randomly, they will break the build, and we will be unable to release new versions or see whether anything has broken. Please fix the tests.

Reply:
The failures are unrelated, but I can't tell what's wrong on my end since the failure only seems to happen on macOS, and test results aren't available on github

Reply:
TypedValueTest > testScriptValidity(ScriptData) > net.sourceforge.kolmafia.textui.parsetree.TypedValueTest.testScriptValidity(ScriptData)[12] FAILED
    org.opentest4j.AssertionFailedError at TypedValueTest.java:121

CustomScriptTest > testScript(String) > net.sourceforge.kolmafia.CustomScriptTest.testScript(String)[3] FAILED
    org.opentest4j.AssertionFailedError at CustomScriptTest.java:67

Above from one of the recent failures.

If you want your commits to continue to be reviewed and submitted a better attitude towards tests might help.

Reply:
at the part exposing how this is actually related to this PR is..?

Both these tests are related to $class[] and $monster[] datatypes, which were updated in r26160 and r26161.

Reply:
I agree that it was presumptuous to phrase it as if it was "definitely" unrelated, but surely you can agree that these errors don't "seem" to come from this PR, right?

Reply:
Your assertion that the failures are unrelated to your changes may be correct but it may reflect a lack of imagination on your part.  Your unwillingness to try and resolve the problem whether or not your code was the direct cause is concerning.

Reply:
No, it was... it was due to the fact that it only seemed to happen on macOS, so I couldn't dig in any further by myself...

BadHorseMonkey helped me with this, so I was able to dig a little further in, and I even think I found a fix, so calling this "unwillingness to try and resolve the problem" kinda hurts... :<

Reply:
Thanks for finding and submitting a fix, fred.

To dig into your earlier statements, while GitHub only allows us to require that the latest attempt at running tests works, in fact we want the tests to pass consistently: so if you ran them three times, say, they would all pass every time. You statement "I finally got them to work" above was not true under this paradigm: you got them to pass on the server once, but not by fixing the problem, instead just running them over and over until they passed once. This isn't great.

I wasn't convinced the failures were unrelated because:
* the failures were all in the parser
* this test modifies something related to the parser
* the tests had passed consistently before this PR

I felt it was more likely this PR had broken something accidentally than a sporadic failure, targeting only parser tests, came out of nowhere during this PR that affects the parser.

I am going to run the tests three times, and if they pass every time, I will merge.